content/silver/concepts/automatic-attributes.md

```diff
@@ -159,11 +162,229 @@ top::Stmt ::= lifted::Decls
 ```
 
 A functor attribute is implemented as just an ordinary synthesized attribute whose type is the same as the type of the nonterminal on which it occurs.  To enable this, a functor attribute forwards to an ordinary synthesized attribute with a type parameter `a`.
-Functor attributes provide an overload for attribute occurrence such that `occurs on` for a functor attribute with no type argument provided will forward to will forward to an attribute occurrence with the nonterminal provided as the type argument.
+Functor attributes provide an overload for attribute occurrence such that `occurs on` for a functor attribute with no type argument provided will forward to an attribute occurrence with the nonterminal provided as the type argument.
 
 `propagate` is overloaded for functor attributes such that propagating a functor attribute will result in an equation that constructs the same production with the result of accessing the attribute on all children.
 Any children on which the attribute does not occur are simply used unchanged in the new tree.
 
+# Destruct attributes
+Destruct attributes can be thought of as sort of an inverse of functor attributes; functor attributes are to tree construction as destruct attributes are to tree deconstruction via pattern matching.  A destruct attribute is an inherited attribute on some tree that decomposes another decorated tree of the same shape; it is an error if a destruct attribute is ever demanded from a tree whose parent was constructed by a different production than the one provided in the attribute.
+
+```
+destruct attribute compareTo;
+
+nonterminal Type;
+
+attribute compareTo occurs on Type;
+
+abstract production intType
+top::Type ::=
+{
+  propagate compareTo;
+}
+
+abstract production dataType
+top::Type ::= name::String
+{
+  propagate compareTo;
+}
+
+abstract production fnType
+top::Type ::= a::Type b::Type
+{
+  propagate compareTo;
+}
+
+```
+
+A destruct attribute might be used for comparing types in a functional language implementation.  The above translates to the following equivalent specification:
```

Comment:
Maybe mention that this example is for types in a functional language before giving the specification. And then say what, for example,`fnType` means.  I assume it is a function type but then names like `a` and `b` aren't so helpful.

Reply:
Maybe I'm looking in the wrong place, but I still see `a` and `b` as names.  Why not `inputType` and `outputType`? These make it very clear what the intended meaning is.

Reply:
I renamed them to `arg` and `ret` but forgot to change it in one place.  I can call then `inputType` and `outputType`.

Reply:
Is there some reason that I still see rhs names of `a` and `b` for `fnType` here in my "review" view of this, but when I go to the repo, the most recent commit has `inputType` and `outputType`?